### ISSUE TITLE
why the loss is so strange?

### ISSUE THREAD

**Issue by HuangZhanPeng** [facebookresearch/StarSpace#145]
here is my train parameters:

```
  -initRandSd 0.01 \
  -ngrams 2 \
  -lr 0.01 \
  -epoch 5 \
  -dim 256 \
  -maxNegSamples 100 \
  -negSearchLimit 100 \
  -trainMode 1 \
  -label "B" \
  -verbose true \
  -thread 24
```

while use adagrad, whatever use any times of epoch, the best loss is the middle epoch, and after that epoch, the loss will arise, for example, total epoch is 10, the best loss will be 4 or 5 epoch, and the loss will begin to arise at the 5 or 6 epoch,like following:

```
Training epoch 0: 0.01 0.002
Epoch: 100.0%  lr: 0.008006  loss: 0.048233  eta: 0h19m  tot: 0h4m53s  (20.0%)
 ---+++                Epoch    0 Train error : 0.04832117 +++--- ☃
Training epoch 1: 0.008 0.002
Epoch: 100.0%  lr: 0.006011  loss: 0.045691  eta: 0h14m  tot: 0h9m42s  (40.0%)
 ---+++                Epoch    1 Train error : 0.04562588 +++--- ☃
Training epoch 2: 0.006 0.002
Epoch: 100.0%  lr: 0.004044  loss: 0.045651  eta: 0h9m  tot: 0h14m33s  (60.0%)
 ---+++                Epoch    2 Train error : 0.04575229 +++--- ☃
Training epoch 3: 0.004 0.002
Epoch: 100.0%  lr: 0.002000  loss: 0.047231  eta: 0h4m  tot: 0h19m34s  (80.0%)
 ---+++                Epoch    3 Train error : 0.04737160 +++--- ☃
Training epoch 4: 0.002 0.002
Epoch: 100.0%  lr: 0.000027  loss: 0.048544  eta: <1min   tot: 0h24m13s  (100.0%)
 ---+++                Epoch    4 Train error : 0.04854773 +++--- ☃
```

and is there any other optimization method to choose, such as sgd, adam .etc.

**Comment by ledw** [comment]
Hi @HuangZhanPeng: depends on your dataset and the learning rate, the loss might go up. You can try to reduce the learning rate to prevent things like overfitting.

**Comment by HuangZhanPeng** [comment]
@ledw I also have a question that, StarSpace how to early stop and save the best model?
sometimes that the last epoch is not the lowest loss

**Comment by ledw** [comment]
@HuangZhanPeng You can set the  -saveEveryEpoch flag to be true, so the model will be saved for every epoch. Early stop is not currently implemented, but is on our plan.

**Comment by HuangZhanPeng** [comment]
@ledw yes, I have try that, but when the model is too large, use -saveEveryEpoch flag will take a very long training time. So looking forward to the new update.
Thanks for answering my question.
